src/action_tools/usage.py

```diff
@@ -0,0 +1,88 @@
+import re
+from typing import Optional
+
+import click
+
+from . import github
+from .models import Action, Resource, Workflow
+
+WORKFLOW_REGEX = re.compile(
+    r"^(?P<org>[\w.-]+)/(?P<repo>[\w.-]+)(?P<subpath>/\.github/workflows/.+\.(yaml|yml))(?:@.+)?$"
+)
+
+
+ACTION_REGEX = re.compile(
+    r"^(?P<org>[\w.-]+)/(?P<repo>[\w.-]+)(?P<subpath>(?:/[\w.-]+)*)(?:@.+)?$"
+)
+
+
+def classify_target(target: str) -> Resource:
+    if match := WORKFLOW_REGEX.match(target):
+        return Workflow(
+            org=match.group("org"),
+            repo=match.group("repo"),
+            subpath=match.group("subpath"),
+        )
+    if match := ACTION_REGEX.match(target):
+        return Action(
+            org=match.group("org"),
+            repo=match.group("repo"),
+            subpath=match.group("subpath"),
+        )
+    raise ValueError(f"target {target} does not appear to be a valid resource")
+
+
+def validate_exists(resource: Resource, client: github.Client) -> bool:
+    try:
+        contents = client.get_repo_contents(
+            org=resource.org, repo=resource.repo, subpath=resource.subpath
+        )
+    except github.ClientError as exc:
+        if exc.response.status_code == 404:
+            return False
+
+    if isinstance(resource, Workflow):
+        return bool(contents)
+    elif isinstance(resource, Action):
+        return any(file["name"] in {"action.yml", "action.yaml"} for file in contents)
+    else:
+        raise ValueError(
+            f"invalid resource: {resource} does not appear to be a path to an action or workflow"
+        )
+
+
+def find_usage(target: str, client: github.Client):
+    query = f'"uses: {target}" language:YAML'
+    items = client.search_code(query)
+    repos = {item["repository"]["full_name"] for item in items}
+    return sorted(repos)
+
+
+def _usage(target: str, token: str, client: Optional[github.Client] = None):
+    if not client:
+        client = github.Client(token=token)
```

Comment:
This is maybe a little weird, by intent is for this to be a way to pass a client for the purposes of DI.